src/kvstore/RocksEngine.cpp

```diff
@@ -19,15 +19,90 @@ using fs::FileType;
 
 const char* kSystemParts = "__system__parts__";
 
+namespace {
+
+/***************************************
+ *
+ * Implementation of WriteBatch
+ *
+ **************************************/
+class RocksWriteBatch : public WriteBatch {
+private:
+    rocksdb::WriteBatch batch_;
+    rocksdb::DB* db_{nullptr};
+
+public:
+    explicit RocksWriteBatch(rocksdb::DB* db) : db_(db) {}
+
+    virtual ~RocksWriteBatch() = default;
+
+    ResultCode put(folly::StringPiece key, folly::StringPiece value) override {
+        if (batch_.Put(toSlice(key), toSlice(value)).ok()) {
+            return ResultCode::SUCCEEDED;
+        } else {
+            return ResultCode::ERR_UNKNOWN;
+        }
+    }
+
+    ResultCode remove(folly::StringPiece key) override {
+        if (batch_.Delete(toSlice(key)).ok()) {
+            return ResultCode::SUCCEEDED;
+        } else {
+            return ResultCode::ERR_UNKNOWN;
+        }
+    }
+
+    ResultCode removePrefix(folly::StringPiece prefix) override {
+        rocksdb::Slice pre(prefix.begin(), prefix.size());
+        rocksdb::ReadOptions options;
+        std::unique_ptr<rocksdb::Iterator> iter(db_->NewIterator(options));
+        iter->Seek(pre);
+        while (iter->Valid()) {
+            if (iter->key().starts_with(pre)) {
+                if (!batch_.Delete(iter->key()).ok()) {
+                    return ResultCode::ERR_UNKNOWN;
+                }
+            } else {
+                // Done
+                break;
+            }
+            iter->Next();
+        }
+        return ResultCode::SUCCEEDED;
+    }
+
+    // Remove all keys in the range [start, end)
+    ResultCode removeRange(folly::StringPiece start, folly::StringPiece end) override {
+        if (batch_.DeleteRange(toSlice(start), toSlice(end)).ok()) {
+            return ResultCode::SUCCEEDED;
+        } else {
+            return ResultCode::ERR_UNKNOWN;
+        }
+    }
+
+    rocksdb::WriteBatch* data() {
+        return &batch_;
+    }
+};
+
+}  // Anonymous namespace
+
+
+/***************************************
+ *
+ * Implementation of WriteBatch
+ *
+ **************************************/
 RocksEngine::RocksEngine(GraphSpaceID spaceId,
                          const std::string& dataPath,
                          std::shared_ptr<rocksdb::MergeOperator> mergeOp,
                          std::shared_ptr<rocksdb::CompactionFilterFactory> cfFactory)
         : KVEngine(spaceId)
-        , dataPath_(dataPath) {
-    LOG(INFO) << "open rocksdb on " << dataPath;
-    if (FileUtils::fileType(dataPath.c_str()) == FileType::NOTEXIST) {
-        FileUtils::makeDir(dataPath);
+        , dataPath_(folly::stringPrintf("%s/nebula/%d", dataPath.c_str(), spaceId)) {
+    auto path = folly::stringPrintf("%s/data", dataPath_.c_str());
+    LOG(INFO) << "open rocksdb on " << path;
+    if (FileUtils::fileType(path.c_str()) == FileType::NOTEXIST) {
```

Comment:
We don't need it.  Rocksdb could create the dir if not exist.

Reply:
Rocksdb will not create the path. It only creates the necessary files

Reply:
Not exactly!   Rocksdb will create the path if not exist,   you could test it.  
But if the parent path not exist,  rocksdb will create path failed.

Reply:
That's exactly the reason here. We need to create the parent path

Reply:
As your code,  dataPath_ is the parent dir.   Fine,  open another issue to track it if you don't want to change in this pr.